### PR TITLE
Retry fed-svc creation on diff NodePort during e2e tests

### LINKED ISSUE #44018: Federated tests should retry NodePort allocation
In order to support running tests in parallel, we added randomized NodePort selection.  There appears to be some evidence that this is not sufficient and is causing e2e test flakes:

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-federation/2510

```
I0331 20:31:25.848] I0401 03:24:49.018147       1 servicecontroller.go:580] Service 'e2e-tests-federated-ingress-fpnqr/federated-ingress-service' is not found in cluster federation-e2e-gce-us-central1-a, trying to create new
I0331 20:31:25.849] I0401 03:24:49.025434       1 servicecontroller.go:587] Failed to create Service "federated-ingress-service" is invalid: spec.ports[0].nodePort: Invalid value: 32589: provided port is already allocated
```

### PR DESCRIPTION
**Pull request by clamoriniere**
**What this PR does / why we need it**:
Currently in federated end2end tests, the creation of services are
done with a randomize NodePort selection take is causing e2e test
flakes if the creation of a federated service failed if the port is
not available.
Now the util.CreateService(...) function is retrying to create the
service on different nodePort in case of error. The method retry until
success or all possible NodePorts have been tested and also failed.

**Which issue this PR fixes** 
fixes #44018

### CODE REVIEW

**Comment by cedriclam** [comment]
CLA signed

File: test/e2e_federation/util.go [hunk 1]
```diff
@@ -150,28 +139,82 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
...
+		i := rand.Intn(len(availablePorts))
+
+		service, err := createServiceWithNodePort(clientset, namespace, name, availablePorts[i])
+		if err == nil {
+			return service, nil
+		}
+
+		// check if service have been created in some clusters.
+		// if it's the case, delete them.
+		if found, _ := hasService(clientset, namespace, name); found {
```
Comment (shashidharatd) [inline test/e2e_federation/util.go]: Here clientset is for federation api server. I think what you wanted was to check if service is created in all federated clusters with the given NodePort and if the service creation in any one of federated clusters has failed, then probably delete that service in federation and ensure that the service is deleted in all federated cluster and create a service with another NodePort. And probably keep doing this until the service is created in all federated clusters with the chosen NodePort.

File: test/e2e_federation/util.go [hunk 2]
```diff
@@ -150,28 +139,82 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
...
-func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, serviceName string, orphanDependents *bool) {
-	if clientset == nil || len(namespace) == 0 || len(serviceName) == 0 {
-		Fail(fmt.Sprintf("Internal error: invalid parameters passed to deleteServiceOrFail: clientset: %v, namespace: %v, service: %v", clientset, namespace, serviceName))
+func hasService(clientset *fedclientset.Clientset, namespace string, serviceName string) (bool, error) {
+	_, err := clientset.Core().Services(namespace).Get(serviceName, metav1.GetOptions{})
+	if err != nil && errors.IsNotFound(err) {
+		return false, err
 	}
-	framework.Logf("Deleting service %q in namespace %v", serviceName, namespace)
+	return true, err
```
Comment (shashidharatd) [inline test/e2e_federation/util.go]: we are returning (true, err) in case of any other errors than `IsNotFound`, which is incorrect. So please put one additional condition to check other errors.

File: test/e2e_federation/util.go [hunk 3]
```diff
@@ -150,28 +139,82 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
...
+	}
 	// Wait for the service to be deleted.
 	err = wait.Poll(5*time.Second, fedframework.FederatedDefaultTestTimeout, func() (bool, error) {
-		_, err := clientset.Core().Services(namespace).Get(serviceName, metav1.GetOptions{})
-		if err != nil && errors.IsNotFound(err) {
-			return true, nil
-		}
-		return false, err
+		found, err := hasService(clientset, namespace, serviceName)
+		return !found, err
```
Comment (shashidharatd) [inline test/e2e_federation/util.go]: prefer to check err here.

**Comment by shashidharatd** [comment]
@k8s-bot ok to test

**Comment by cedriclam** [comment]
@shashidharatd thanks for your review. I updated the PR with your remarks.

File: test/e2e_federation/util.go [hunk 4]
```diff
@@ -150,28 +139,104 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
...
 	err := clientset.Services(namespace).Delete(serviceName, &metav1.DeleteOptions{OrphanDependents: orphanDependents})
-	framework.ExpectNoError(err, "Error deleting service %q from namespace %q", serviceName, namespace)
+	if err != nil {
+		return err
+	}
 	// Wait for the service to be deleted.
 	err = wait.Poll(5*time.Second, fedframework.FederatedDefaultTestTimeout, func() (bool, error) {
-		_, err := clientset.Core().Services(namespace).Get(serviceName, metav1.GetOptions{})
-		if err != nil && errors.IsNotFound(err) {
+		found, err := hasService(clientset, namespace, serviceName)
```
Comment (shashidharatd) [inline test/e2e_federation/util.go]: I don't find any significant value in hasService function and it has made the logic little complex. Can we get rid of this function and use `clientset.Core().Services(namespace).Get(serviceName, metav1.GetOptions{})` as earlier.

Comment (cedriclam) [inline test/e2e_federation/util.go]: right

File: test/e2e_federation/util.go [hunk 5]
```diff
@@ -150,20 +139,78 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
...
+		if err == nil {
+			return service, nil
+		}
+
+		// check if service have been created in some clusters.
+		// if it's the case, delete them.
+		needsToBeDeleted := false
+		err = wait.Poll(5*time.Second, fedframework.FederatedDefaultTestTimeout, func() (bool, error) {
+			var err error
+			_, err = clientset.Core().Services(namespace).Get(name, metav1.GetOptions{})
```
Comment (shashidharatd) [inline test/e2e_federation/util.go]: Additionally we should be checking for service shard creation in federated cluster. Refer to second point in @madhusudancs 's suggestion in https://github.com/kubernetes/kubernetes/issues/44018#issuecomment-298218566

Comment (cedriclam) [inline test/e2e_federation/util.go]: Your are right, 
I was thinking that calling Services(...).Delete(...) with the fedclientset.Clientset, will cascade the deletion for service shard.
I will add a loop on the ClusterSlice to call the cleanupServiceShard() function

**Comment by shashidharatd** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by shashidharatd** [comment]
@cedriclam, sorry for the delay in checking this PR. I find some issues in the logic. PTAL at comments.

---

Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[test/e2e_federation/util.go, line 172 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-Kj4svaqgMvKvZ3-Ez7P-r3-172:-Kjvc5WuUZq1ziO-qsRA:bpualp) ([raw file](https://github.com/kubernetes/kubernetes/blob/6abb510985f6ce62df80534de779fa1c7967f218/test/e2e_federation/util.go#L172)):*
<details><summary><i>Previously, cedriclam (Cedric Lamoriniere) wrote…</i></summary><blockquote>

Your are right, 
I was thinking that calling Services(...).Delete(...) with the fedclientset.Clientset, will cascade the deletion for service shard.
I will add a loop on the ClusterSlice to call the cleanupServiceShard() function
</blockquote></details>

Here in wait.Poll, you are checking if service is created in federation only. We should instead be checking for service shard creation in all federated clusters (in a loop for federated clusters. even if one federated cluster failed to create service, we should cleanup and try with anoter NodePort).

---

*[test/e2e_federation/util.go, line 164 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-KjvbRQVnAM9fIzA7DJ3:-KjvbRQVnAM9fIzA7DJ4:b-qpyhae) ([raw file](https://github.com/kubernetes/kubernetes/blob/264017c14475c06d21a5b10fc24fc38fc225fec2/test/e2e_federation/util.go#L164)):*
> ```Go
> 		service, err := createServiceWithNodePort(clientset, namespace, name, availablePorts[i])
> 		if err == nil {
> 			return service, nil
> ```

We should not return from here. We are just successful in creating a NodePort service in federation. we still have to ensure all federated clusters are able to create the service with the same NodePort.

---

*[test/e2e_federation/util.go, line 233 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-Kjv_rKBcGw9vMwEEX1C:-Kjv_rKCTBp4GoD4-x8G:b-dqv32h) ([raw file](https://github.com/kubernetes/kubernetes/blob/264017c14475c06d21a5b10fc24fc38fc225fec2/test/e2e_federation/util.go#L233)):*
> ```Go
> 	if err != nil {
> 		framework.DescribeSvc(namespace)
> 		framework.Failf("Error in deleting service %s: %v", serviceName, err)
> ```

Log the serviceName along with namespace, so that its easier to debug in case of failures.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-:-KjvcqlKrTS8TPZi4Xq1:beun67y)*

**Comment by cedriclam** [comment]
@shashidharatd thanks for your review, I will update the PR with your comments this evenings.

**Comment by cedriclam** [comment]
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[test/e2e_federation/util.go, line 169 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-Kj4svaqgMvKvZ3-Ez7P-r1-169:-KkSMMChylYRN_G2QhHD:bb7f2mf) ([raw file](https://github.com/kubernetes/kubernetes/blob/1ea4860ec159959bba5ca606ca2d700e416a7fdd/test/e2e_federation/util.go#L169)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

Here clientset is for federation api server. I think what you wanted was to check if service is created in all federated clusters with the given NodePort and if the service creation in any one of federated clusters has failed, then probably delete that service in federation and ensure that the service is deleted in all federated cluster and create a service with another NodePort. And probably keep doing this until the service is created in all federated clusters with the chosen NodePort.
</blockquote></details>

done

---

*[test/e2e_federation/util.go, line 172 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-Kj4svaqgMvKvZ3-Ez7P-r3-172:-KkSMUQRW9cDo1OWkZdb:bda1wk) ([raw file](https://github.com/kubernetes/kubernetes/blob/6abb510985f6ce62df80534de779fa1c7967f218/test/e2e_federation/util.go#L172)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

Here in wait.Poll, you are checking if service is created in federation only. We should instead be checking for service shard creation in all federated clusters (in a loop for federated clusters. even if one federated cluster failed to create service, we should cleanup and try with anoter NodePort).
</blockquote></details>

now the checkServicesCreation(...) is checking in all federated clusters the creation of the service.

---

*[test/e2e_federation/util.go, line 164 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-KjvbRQVnAM9fIzA7DJ3:-KkSLGFSjelYPWAa6tj9:b-c4bmsp) ([raw file](https://github.com/kubernetes/kubernetes/blob/264017c14475c06d21a5b10fc24fc38fc225fec2/test/e2e_federation/util.go#L164)):*
> @shashidharatd right, code updated

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186)*

**Comment by cedriclam** [comment]
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[test/e2e_federation/util.go, line 196 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-Kj4svaqgMvKvZ3-Ez7P-r1-196:-KkSYBGlI3pIrh2s0iO-:b-qssecl) ([raw file](https://github.com/kubernetes/kubernetes/blob/1ea4860ec159959bba5ca606ca2d700e416a7fdd/test/e2e_federation/util.go#L196)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

we are returning (true, err) in case of any other errors than `IsNotFound`, which is incorrect. So please put one additional condition to check other errors.
</blockquote></details>

fixed

---

*[test/e2e_federation/util.go, line 207 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-Kj4svaqgMvKvZ3-Ez7P-r1-207:-KkSOiZg10R5OJZD1vhx:b-7ql1td) ([raw file](https://github.com/kubernetes/kubernetes/blob/1ea4860ec159959bba5ca606ca2d700e416a7fdd/test/e2e_federation/util.go#L207)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

prefer to check err here.
</blockquote></details>

fixed

---

*[test/e2e_federation/util.go, line 233 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-Kjv_rKBcGw9vMwEEX1C:-KkSP57us_97T2Pe_f88:b-7ql1td) ([raw file](https://github.com/kubernetes/kubernetes/blob/264017c14475c06d21a5b10fc24fc38fc225fec2/test/e2e_federation/util.go#L233)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

Log the serviceName along with namespace, so that its easier to debug in case of failures.
</blockquote></details>

fixed

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186)*

**Comment by shashidharatd** [comment]
@madhusudancs, very strange thing happened in federation presubmit tests. it says, it still running after 6 hours. https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/45186/pull-kubernetes-federation-e2e-gce/5493/

**Comment by shashidharatd** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by shashidharatd** [comment]
@cedriclam, there seems to be some problem with the logic. the tests are not completing. Can you please analyze logs and find the cause?

**Comment by cedriclam** [comment]
@shashidharatd yes sure

File: test/e2e_federation/authn.go [hunk 6]
```diff
@@ -35,16 +35,20 @@ var _ = framework.KubeDescribe("[Feature:Federation]", func() {
 	f := fedframework.NewDefaultFederatedFramework("federation-apiserver-authn")
 
 	var _ = Describe("Federation API server authentication [NoCluster]", func() {
+		var (
+			clusters fedframework.ClusterSlice
+		)
 		BeforeEach(func() {
 			fedframework.SkipUnlessFederated(f.ClientSet)
+			clusters = fedframework.ClusterSlice{}
```
Comment (shashidharatd) [inline test/e2e_federation/authn.go]: here you are initializing the clusters with 0 clusters which is not the correct thing to do. In this case, the validation to see if the service was created in federated cluster would not work as it won't enter the for loop of clusters to check at all.
Instead of passing clusters from every caller. please call
```
clusters = f.GetRegisteredClusters()
```
in createService itself.

File: test/e2e_federation/service.go [hunk 7]
```diff
@@ -102,15 +103,15 @@ var _ = framework.KubeDescribe("Federated Services [Feature:Federation]", func()
 
 				if service != nil {
 					By(fmt.Sprintf("Deleting service shards and their provider resources in underlying clusters for service %q in namespace %q", service.Name, nsName))
-					cleanupServiceShardsAndProviderResources(nsName, service, clusters)
+					cleanupServiceShardsAndProviderResources(nsName, service.Name, clusters)
```
Comment (shashidharatd) [inline test/e2e_federation/service.go]: we could always get `service.Name` for `service` right? why this change is required. Please try to keep your changes to minimal unless there is a strong case in doing so.

File: test/e2e_federation/authn.go (hunk 6 above)
Comment (cedriclam) [inline test/e2e_federation/authn.go]: understood

File: test/e2e_federation/service.go (hunk 7 above)
Comment (cedriclam) [inline test/e2e_federation/service.go]: Ok, I done this modification since the only "service *api.Service" information retrieve in the "cleanupServiceShardsAndProviderResources()" method was its name.

But sure, I will rollback those changes in order to minimalise the PR's footprint .

**Comment by cedriclam** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by cedriclam** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by cedriclam** [comment]
Hi @shashidharatd 
I updated the PR with your last comments, but the e2e_federation test seams to have some initialisation issue... Could you help me on it?

File: test/e2e_federation/authn.go [hunk 8]
```diff
@@ -35,16 +35,20 @@ var _ = framework.KubeDescribe("[Feature:Federation]", func() {
...
+			clusters = f.GetRegisteredClusters()
 		})
 
 		It("should accept cluster resources when the client has certificate authentication credentials", func() {
 			fcs, err := federationClientSetWithCert()
 			framework.ExpectNoError(err)
 
 			nsName := f.FederationNamespace.Name
-			svc := createServiceOrFail(fcs, nsName, FederatedServiceName)
+			svc := createServiceOrFail(fcs, nsName, FederatedServiceName, clusters)
```
Comment (shashidharatd) [inline test/e2e_federation/authn.go]: @cedriclam, can you please revert passing the clusters from caller of `createServiceOrFail` and call `clusters = f.GetRegisteredClusters()` inside `createServiceOrFail` function.

File: test/e2e_federation/util.go [hunk 9]
```diff
@@ -172,33 +226,57 @@ func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, se
...
+		cSvc, err = cluster.Clientset.Services(namespace).Get(serviceName, metav1.GetOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			// Get failed with an error, try again.
+			framework.Logf("Failed to find service %q in namespace %q, in cluster %q: %v. Trying again in %s", serviceName, namespace, name, err, framework.Poll)
+			return false, nil
+		} else if errors.IsNotFound(err) {
+			cSvc = nil
+			if retry {
+				// don't return the
+				framework.Logf("Service %q in namespace %q in cluster %q not found. Trying again in %s", serviceName, namespace, name, framework.Poll)
```
Comment (shashidharatd) [inline test/e2e_federation/util.go]: broken comment

File: test/e2e_federation/util.go [hunk 10]
```diff
@@ -172,33 +226,57 @@ func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, se
...
+				return false, nil
+			}
+			By(fmt.Sprintf("Service %q in namespace %q in cluster %q not found", serviceName, namespace, name))
+			return true, err
+		}
+		By(fmt.Sprintf("Service %q in namespace %q in cluster %q found", serviceName, namespace, name))
+		return true, err
```
Comment (shashidharatd) [inline test/e2e_federation/util.go]: shouldn't this be return true, nil?

File: test/e2e_federation/authn.go (hunk 8 above)
Comment (cedriclam) [inline test/e2e_federation/authn.go]: in order to call 'clusters = f.GetRegisteredClusters()' inside `createServiceOrFail`, I need to access the instance of `f *framework.Framework`. 
Also `createServiceOrFail` is a generic function used in deferent test cases, that initialise differently the  `f *framework.Framework`. 

@shashidharatd, Do you prefer that it stay like it is, or change the function signature by replacing `clusters` by `f` ? or do you see another way to revert passing `clusters` from caller of `createServiceOrFail`.

File: test/e2e_federation/util.go (hunk 10 above)
Comment (cedriclam) [inline test/e2e_federation/util.go]: right.

File: test/e2e_federation/util.go (hunk 9 above)
Comment (cedriclam) [inline test/e2e_federation/util.go]: comment update

**Comment by shashidharatd** [comment]
@cedriclam, I am very sorry for the delay got tied up in some other priority work. will have a final look today.
Also request to squash all the commits.

**Comment by cedriclam** [comment]
@shashidharatd, no problem for the delay, I completely understand that you can have other priority work :).

I squashed all the commits.

**Comment by cedriclam** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by cedriclam** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

File: test/e2e_federation/authn.go (hunk 8 above)
Comment (shashidharatd) [inline test/e2e_federation/authn.go]: OK, i understand your predicament. sorry for that. its fine to pass clusters. better than the alternative

File: test/e2e_federation/service.go [hunk 11]
```diff
@@ -53,6 +53,7 @@ var _ = framework.KubeDescribe("Federated Services [Feature:Federation]", func()
 	var _ = Describe("Without Clusters [NoCluster]", func() {
 		BeforeEach(func() {
 			fedframework.SkipUnlessFederated(f.ClientSet)
+			clusters = fedframework.ClusterSlice{}
```
Comment (shashidharatd) [inline test/e2e_federation/service.go]: we should not do this. as my earlier comment. we should get actual clusters. like `clusters = f.GetRegisteredClusters()`

File: test/e2e_federation/util.go [hunk 12]
```diff
@@ -150,20 +139,85 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
...
+		if err == nil {
+			// check if service have been created properly in all clusters.
+			// if the service is not present in one of the clusters, we should cleanup all services
+			err = checkServicesCreation(namespace, name, clusters)
+		}
+
+		if err == nil {
+			// everything was created properly so returns the federated service.
+			return service, nil
+		}
```
Comment (shashidharatd) [inline test/e2e_federation/util.go]: This block should be merged with previous block. why two `if err == nil `?

File: test/e2e_federation/util.go [hunk 13]
```diff
@@ -150,20 +139,85 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
...
+
+		// in case of error, cleanup everything
+		if service != nil {
+			if err = deleteService(clientset, namespace, name, nil); err != nil {
```
Comment (shashidharatd) [inline test/e2e_federation/util.go]: why deleteService in just one cluster, when you are calling `cleanupServiceShardsAndProviderResources` 4 lines below which does cleanup all cluster services.

File: test/e2e_federation/util.go [hunk 14]
```diff
@@ -150,20 +139,85 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
...
+		return nil, fmt.Errorf("Internal error: invalid parameters passed to createService: clientset: %v, namespace: %v", clientset, namespace)
+	}
+	By(fmt.Sprintf("Creating federated service %q in namespace %q", name, namespace))
+
+	// Tests can be run in parallel, so we need a different nodePort for
+	// each test.
+	// we add in a array all the "available" ports
+	availablePorts := make([]int32, FederatedSvcNodePortLast-FederatedSvcNodePortFirst)
+	for i := range availablePorts {
+		availablePorts[i] = int32(FederatedSvcNodePortFirst + i)
```
Comment (shashidharatd) [inline test/e2e_federation/util.go]: This list of available ports is kind of very big. in case of some network error to some cluster. we would keep on retrying forever in the loop. So i would like it to limit to 5-10 iterations at max.

File: test/e2e_federation/util.go [hunk 15]
```diff
@@ -172,33 +226,57 @@ func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, se
...
+			// Get failed with an error, try again.
+			framework.Logf("Failed to find service %q in namespace %q, in cluster %q: %v. Trying again in %s", serviceName, namespace, name, err, framework.Poll)
+			return false, nil
+		} else if errors.IsNotFound(err) {
+			cSvc = nil
+			if retry {
+				// the function returns false in order to test again until timeout.
+				framework.Logf("Service %q in namespace %q in cluster %q not found. Trying again in %s", serviceName, namespace, name, framework.Poll)
+				return false, nil
+			}
```
Comment (shashidharatd) [inline test/e2e_federation/util.go]: introducing another parameter retry and adding a if condition is increasing the complexity of the function anyway. So i would suggest to not merge multiple scenarios into the same wait.PollImmediate function.

better to leave the original `cleanupServiceShardsAndProviderResources` intact. and use another `wait.PollImmediate` function for the other scenario.

File: test/e2e_federation/util.go (hunk 13 above)
Comment (cedriclam) [inline test/e2e_federation/util.go]: I was thinking that the `clientset` is the client to target the Federation-API-server, and so not one of the federated cluster. 
That why I wanted first to delete the Service at federation level. then iterate on all the federated clusters.

File: test/e2e_federation/util.go (hunk 12 above)
Comment (cedriclam) [inline test/e2e_federation/util.go]: first`ìf err == nil` is here for validation the the Service was created properly in the Federation-API-Server
the second `ìf err == nil` is here to check that the service have been creation on each federated clusters.

I didn't initially merge both, because I was considering that if `createServiceWithNodePort()` already return an error, we don't need to check on each cluster if the service was created properly.

File: test/e2e_federation/util.go (hunk 15 above)
Comment (cedriclam) [inline test/e2e_federation/util.go]: understood

File: test/e2e_federation/service.go (hunk 11 above)
Comment (cedriclam) [inline test/e2e_federation/service.go]: right, I forgot to make this modification that you already spot in a previous comment.

File: test/e2e_federation/util.go (hunk 14 above)
Comment (cedriclam) [inline test/e2e_federation/util.go]: understood, I will add a `retry`counter in order to limit the number of iterations.

**Comment by cedriclam** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by cedriclam** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

File: test/e2e_federation/util.go [hunk 16]
```diff
@@ -172,33 +226,70 @@ func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, se
...
-				return false, nil
-			} else if errors.IsNotFound(err) {
-				cSvc = nil
-				By(fmt.Sprintf("Service %q in namespace %q in cluster %q not found", service.Name, namespace, name))
-				return true, err
-			}
-			By(fmt.Sprintf("Service %q in namespace %q in cluster %q found", service.Name, namespace, name))
-			return true, err
-		})
+		cSvc, err := getServiceShard(namespace, service.Name, c)
```
Comment (shashidharatd) [inline test/e2e_federation/util.go]: @cedriclam  There was no need of pulling out code and creating `getServiceShard` function. Can you please revert it. I would apply the LGTM after this change. Thanks !

File: test/e2e_federation/util.go [hunk 17]
```diff
@@ -172,33 +226,70 @@ func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, se
...
+			return true, err
+		}
+		By(fmt.Sprintf("Service %q in namespace %q in cluster %q found", serviceName, namespace, name))
+		return true, nil
+	})
+
+	return cSvc, err
+}
+
+func getServiceShardWithRetry(namespace string, serviceName string, cluster *fedframework.Cluster) (*v1.Service, error) {
```
Comment (shashidharatd) [inline test/e2e_federation/util.go]: Also please integrate this function `getServiceShardWithRetry` in `checkServicesCreation` itself.

File: test/e2e_federation/util.go [hunk 18]
```diff
@@ -172,33 +226,70 @@ func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, se
...
+	name := cluster.Name
+	var cSvc *v1.Service
+	err := wait.PollImmediate(framework.Poll, fedframework.FederatedDefaultTestTimeout, func() (bool, error) {
+		var err error
+		cSvc, err = cluster.Clientset.Services(namespace).Get(serviceName, metav1.GetOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			// Get failed with an error, try again.
+			framework.Logf("Failed to find service %q in namespace %q, in cluster %q: %v. Trying again in %s", serviceName, namespace, name, err, framework.Poll)
+			return false, nil
```
Comment (shashidharatd) [inline test/e2e_federation/util.go]: it would be better to `return false, err` here, if any err other than notfound err. Since it may be non-recoverable error and there is no point in keep on retrying.

Comment (cedriclam) [inline test/e2e_federation/util.go]: right, thanks for pointing it out.

File: test/e2e_federation/util.go (hunk 16 above)
Comment (cedriclam) [inline test/e2e_federation/util.go]: ok cool!

File: test/e2e_federation/util.go (hunk 17 above)
Comment (cedriclam) [inline test/e2e_federation/util.go]: understood

**Comment by cedriclam** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Comment by nikhiljindal** [comment]
/release-note-none

/approve.

Will let @shashidharatd add LGTM

File: test/e2e_federation/util.go [hunk 19]
```diff
@@ -150,20 +139,101 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
...
+	for _, cluster := range clusters {
+		name := cluster.Name
+		var cSvc *v1.Service
+		err := wait.PollImmediate(framework.Poll, fedframework.FederatedDefaultTestTimeout, func() (bool, error) {
+			var err error
+			cSvc, err = cluster.Clientset.Services(namespace).Get(serviceName, metav1.GetOptions{})
+			if err != nil && !errors.IsNotFound(err) {
+				// Get failed with an error, try again.
+				framework.Logf("Failed to find service %q in namespace %q, in cluster %q: %v. Trying again in %s", serviceName, namespace, name, err, framework.Poll)
+				return false, nil
```
Comment (shashidharatd) [inline test/e2e_federation/util.go]: @cedriclam, just the last nit. Better to return false,   err in this sentence, as any other error could potentially be non-recoverable and there is no point in waiting and retrying.

**Comment by shashidharatd** [comment]
@cedriclam, Sorry that you had to rebase. We had to do some optimizations in the tests.
Can you please rebase this PR and address the conflicts. We could still be potentially include this PR in v1.7 as this is a test-only change.

**Comment by cedriclam** [comment]
@shashidharatd, PR rebase done
let me know if I need to change other stuff.

**Comment by cedriclam** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by cedriclam** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by cedriclam** [comment]
@shashidharatd mind have a look to this PR? sorry for the late PR fix after rebase.

**Comment by cedriclam** [comment]
@madhusudancs mind have a look at this PR?

File: test/e2e_federation/util.go [hunk 20]
```diff
@@ -210,18 +310,14 @@ func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, se
...
-	}
+	return err
 }
 
 func cleanupServiceShardsAndProviderResources(namespace string, service *v1.Service, clusters fedframework.ClusterSlice) {
 	framework.Logf("Deleting service %q in %d clusters", service.Name, len(clusters))
 	for _, c := range clusters {
 		name := c.Name
 		var cSvc *v1.Service
-
```
Comment (shashidharatd) [inline test/e2e_federation/util.go]: unnecessary change

File: test/e2e_federation/util.go [hunk 21]
```diff
@@ -141,21 +141,84 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
 	}
 
 	By(fmt.Sprintf("Trying to create service %q in namespace %q", service.Name, namespace))
-	return clientset.Services(namespace).Create(service)
+	var err error
```
Comment (shashidharatd) [inline test/e2e_federation/util.go]: Given that we have 2 functions now `createService` and `createLBService`, checking and deleting if service shard is created in all member cluster in case of simple service is a overkill. I think the logic to check successful creation of service shard and retrying should be good enough in the scenario of `createLBService`.

File: test/e2e_federation/util.go [hunk 22]
```diff
@@ -237,7 +333,6 @@ func cleanupServiceShardsAndProviderResources(namespace string, service *v1.Serv
 			By(fmt.Sprintf("Service %q in namespace %q in cluster %q found", service.Name, namespace, name))
 			return true, err
 		})
-
```
Comment (shashidharatd) [inline test/e2e_federation/util.go]: unnecessary change

**Review by shashidharatd — CHANGES_REQUESTED** [review]
@cedriclam, I am extremely sorry for such a long delay, as i was not well and had taken a break. I wish we could address the comments in this PR and close it within couple of days.
The concern i have is to minimize the test execution time in case of failures and would like to touch minimal code to make these changes.
Thanks a lot for holding up.

File: test/e2e_federation/util.go (hunk 21 above)
Comment (cedriclam) [inline test/e2e_federation/util.go]: I updated the createService function as you proposed: create federated service, then check creation.

**Comment by cedriclam** [comment]
@shashidharatd now worries for the delay. I updated the PR with your comments.
Let me know if it is looking good for you, and if you see other changes needed.
Thanks for your time.

File: test/e2e_federation/util.go [hunk 23]
```diff
@@ -141,21 +141,71 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
 	}
 
 	By(fmt.Sprintf("Trying to create service %q in namespace %q", service.Name, namespace))
-	return clientset.CoreV1().Services(namespace).Create(service)
+	var err error
```
Comment (shashidharatd) [inline test/e2e_federation/util.go]: This block may not be required and the original one should suffice. Can you please revert this one and also preserve the signature of createService function

Comment (cedriclam) [inline test/e2e_federation/util.go]: sure
